### PR TITLE
Whitelist otoken

### PR DESCRIPTION
**Pull request by haythem96**
This PR add the feature of whitelisting a specific `otoken` by the `Otoken Factory`, related to the `Whitelist` Module.

### CODE REVIEW

File: contracts/Whitelist.sol [hunk 1]
```diff
@@ -1,6 +1,8 @@
 // SPDX-License-Identifier: UNLICENSED
 pragma solidity 0.6.10;
 
+import "./interfaces/AddressBookInterface.sol";
+
```
Comment (aparnakr) [inline contracts/Whitelist.sol:5]: delete the extra empty line

File: contracts/Whitelist.sol [hunk 2]
```diff
@@ -9,8 +11,23 @@ import "./packages/oz/Ownable.sol";
...
+    /**
+     * @dev constructor
+     * @param _addressBook AddressBook module address
+     */
+    constructor(address _addressBook) public {
+        require(_addressBook != address(0), "Invalid address book");
+
+        addressBook = _addressBook;
+    }
```
Comment (aparnakr) [inline contracts/Whitelist.sol:30]: Can anyone call the constructor or just the owner?

File: contracts/Whitelist.sol [hunk 3]
```diff
@@ -20,13 +37,25 @@ contract Whitelist is Ownable {
         address indexed collateral
     );
 
+    ///@notice emitted when Otoken Factory module whitelist an otoken
```
Comment (aparnakr) [inline contracts/Whitelist.sol:40]: `///@notice emitted when the Otoken Factory module whitelists an otoken`

File: contracts/Whitelist.sol [hunk 4]
```diff
@@ -9,8 +11,23 @@ import "./packages/oz/Ownable.sol";
...
+     * @dev constructor
+     * @param _addressBook AddressBook module address
+     */
+    constructor(address _addressBook) public {
+        require(_addressBook != address(0), "Invalid address book");
+
+        addressBook = _addressBook;
+    }
 
     ///@notice emitted when owner whitelist a product
```
Comment (aparnakr) [inline contracts/Whitelist.sol:32]: `///@notice emitted when the owner whitelists a new product`

File: contracts/Whitelist.sol [hunk 5]
```diff
@@ -9,8 +11,23 @@ import "./packages/oz/Ownable.sol";
  * @notice The whitelist module keeps track of all valid Otoken contracts.
  */
 contract Whitelist is Ownable {
-    ///@notice mapping to track whitelisted product
+    ///@notice AddressBook module address
+    address public addressBook;
+
+    ///@dev mapping to track whitelisted product
```
Comment (aparnakr) [inline contracts/Whitelist.sol:17]: `///@dev mapping to track whitelisted products`

File: contracts/Whitelist.sol [hunk 6]
```diff
@@ -9,8 +11,23 @@ import "./packages/oz/Ownable.sol";
...
     mapping(bytes32 => bool) internal whitelistedProduct;
+    ///@dev mapping to track whitelisted otoken
```
Comment (aparnakr) [inline contracts/Whitelist.sol:19]: `///@dev mapping to track whitelisted otokens`

File: contracts/Whitelist.sol (hunk 2 above)
Comment (haythem96) [reply contracts/Whitelist.sol:30]: constructor called only at deployment time.

File: contracts/Whitelist.sol (hunk 3 above)
Comment (antoncoding) [reply contracts/Whitelist.sol:39]: so when are we using /// @notice ?

File: contracts/Whitelist.sol [hunk 7]
```diff
@@ -9,24 +10,51 @@ import "./packages/oz/Ownable.sol";
...
 
+    ///@notice emitted when Otoken Factory module whitelist an otoken
+    event OtokenWhitelisted(address otoken);
+
     /**
-     * @notice check if a product is supported
+     * @notice check if sender is Otoken Factory module
+     */
+    modifier onlyFactory() {
+        require(msg.sender == AddressBookInterface(addressBook).getOtokenFactory(), "Sender is not Otoken Factory");
```
Comment (antoncoding) [inline contracts/Whitelist.sol:46]: I suggest adding the contract name to the require message, ( `Sender is not Otoken Factory` => `WhiteList: Sender is not Otoken Factory`, `WL:Sender is not Otoken Factory`)  that way it's much easier for us to know which contract is throwing errors while testing.

File: test/Whitelist.test.ts [hunk 8]
```diff
@@ -1,16 +1,21 @@
...
 const {expectEvent, expectRevert} = require('@openzeppelin/test-helpers')
 
 const MockERC20 = artifacts.require('MockERC20.sol')
+const MockAddressBook = artifacts.require('MockAddressBook.sol')
 const Whitelist = artifacts.require('Whitelist.sol')
 // address(0)
 const ZERO_ADDR = '0x0000000000000000000000000000000000000000'
 
-contract('Whitelist', ([owner, random, newOwner]) => {
+contract('Whitelist', ([owner, otokenFactoryAddress, random, newOwner]) => {
```
Comment (antoncoding) [inline test/Whitelist.test.ts:11]: nice, we should use this

File: contracts/Whitelist.sol [hunk 9]
```diff
@@ -59,12 +96,33 @@ contract Whitelist is Ownable {
     }
 
     /**
-     * @notice set a product hash as supported
+     * @notice allow Otoken Factory to whitelist option
+     * @dev can only be called from owner address
+     * @param _otokenAddress otoken
+     */
+    function whitelistOtoken(address _otokenAddress) external onlyFactory {
```
Comment (antoncoding) [inline contracts/Whitelist.sol:103]: Didn't we change the name to `registerOtoken` already?

File: contracts/Whitelist.sol [hunk 10]
```diff
@@ -38,6 +66,15 @@ contract Whitelist is Ownable {
         return whitelistedProduct[productHash];
     }
 
+    /**
+     * @notice check if an otoken is whitelisted
+     * @param _otoken otoken address
+     * @return boolean, true if otoken is whitelisted
+     */
+    function isWhitelistedOtoken(address _otoken) external view returns (bool) {
+        return whitelistedOtoken[_otoken];
```
Comment (antoncoding) [inline contracts/Whitelist.sol:75]: `isValidOtoken` make more sense to me. It's not like we only whitelist several otokens out of all the otokens, we literally whitelist every valid token created by the factory.

File: contracts/Whitelist.sol (hunk 9 above)
Comment (antoncoding) [reply contracts/Whitelist.sol:103]: I saw your comments on the interface thread, but I think we better discuss this more.

File: contracts/Whitelist.sol (hunk 10 above)
Comment (haythem96) [reply contracts/Whitelist.sol:75]: yeah thought better to use same terminology everywhere, "Whitelist module, whitelist, isWhitelisted", even if we whitelist all token, it is still the same as valid as we are marking all token as valid. I think better to use same terminology.

Comment (antoncoding) [reply contracts/Whitelist.sol:75]: Let's use this for now, could easily change it back if we end up having different decision.

**Comment by haythem96** [comment]
> Need to resolve the oToken naming part

What naming part ?

File: contracts/Whitelist.sol (hunk 3 above)
Comment (aparnakr) [reply contracts/Whitelist.sol:39]: for events and variables!

File: contracts/Whitelist.sol [hunk 11]
```diff
@@ -9,24 +10,54 @@ import "./packages/oz/Ownable.sol";
...
         address indexed strike,
         address indexed collateral
     );
 
+    ///@notice emitted when Otoken Factory module whitelist an otoken
+    event OtokenWhitelisted(address otoken);
+
     /**
-     * @notice check if a product is supported
+     * @notice check if sender is Otoken Factory module
```
Comment (aparnakr) [inline contracts/Whitelist.sol:43]:
```suggestion
     * @notice check if the sender is the Otoken Factory module
```

File: contracts/Whitelist.sol [hunk 12]
```diff
@@ -9,24 +10,54 @@ import "./packages/oz/Ownable.sol";
...
+
+        _;
+    }
+
+    /**
+     * @notice check if a product is whitelisted
      * @dev product = the hash of underlying, strike and collateral asset
      * @param _underlying option underlying asset address
      * @param _strike option strike asset address
      * @param _collateral option collateral asset address
```
Comment (aparnakr) [inline contracts/Whitelist.sol:59]: update these with the comment @aleone had on https://github.com/opynfinance/GammaProtocol/pull/6

File: contracts/Whitelist.sol (hunk 10 above)
Comment (aparnakr) [reply contracts/Whitelist.sol:78]: +1 on Anton's comment

File: contracts/Whitelist.sol [hunk 13]
```diff
@@ -59,12 +99,33 @@ contract Whitelist is Ownable {
     }
 
     /**
-     * @notice set a product hash as supported
+     * @notice allow Otoken Factory to whitelist option
+     * @dev can only be called from owner address
```
Comment (aparnakr) [inline contracts/Whitelist.sol:103]:
```suggestion
     * @dev can only be called from the owner's address
```

File: contracts/Whitelist.sol [hunk 14]
```diff
@@ -59,12 +99,33 @@ contract Whitelist is Ownable {
     }
 
     /**
-     * @notice set a product hash as supported
+     * @notice allow Otoken Factory to whitelist option
```
Comment (aparnakr) [inline contracts/Whitelist.sol:102]:
```suggestion
     * @notice allow Otoken Factory to whitelist a new option
```

File: contracts/Whitelist.sol [hunk 15]
```diff
@@ -59,12 +99,33 @@ contract Whitelist is Ownable {
...
      */
     function _setWhitelistedProduct(bytes32 _productHash) internal {
         require(whitelistedProduct[_productHash] == false, "Product already whitelisted");
 
         whitelistedProduct[_productHash] = true;
     }
+
+    /**
+     * @notice set an otoken address as whitelisted
+     * @param _otokenAddress product hash in bytes
```
Comment (aparnakr) [inline contracts/Whitelist.sol:126]:
```suggestion
     * @param _otokenAddress address of the oToken that is being whitelisted
```

File: test/Whitelist.test.ts [hunk 16]
```diff
@@ -50,6 +63,28 @@ contract('Whitelist', ([owner, random, newOwner]) => {
     })
   })
 
+  describe('Whitelist otoken', () => {
+    it('should revert whitelisting an otoken from sender other than Otoken Factory', async () => {
+      await expectRevert(whitelist.whitelistOtoken(otoken.address, {from: owner}), 'Sender is not Otoken Factory')
+    })
+
+    it('should whitelist an otoken from Otoken Factory address', async () => {
```
Comment (aparnakr) [inline test/Whitelist.test.ts:71]: Can we add a test to ensure that after Factory Address has changed in the addressBook, the new oToken can still be whitelisted

**Comment by aparnakr** [comment]
Great work!

File: test/Whitelist.test.ts (hunk 16 above)
Comment (haythem96) [reply test/Whitelist.test.ts:71]: In upgradability, the first deployed address will alway be the caller, as `V1` will delegate call to `V2`, then `V2` will execute, in the execution transaction, `V1` is the `msg.sender`. (V1 is first version of the contract and V2 is an upgrade to a second version).

Comment (haythem96) [reply test/Whitelist.test.ts:71]: Or no I think I'm wrong in this, will make sure of it, if the contract address change when it make an external call, will add the needed test for that.

Comment (haythem96) [reply test/Whitelist.test.ts:71]: Sorry my first response is out-of-topic, so when we change the implementation address of the Factory module in the addressBook, nothing change for the whitelist module as the `onlyFactory` modifier always call the addressBook to get the updated implementation address.

**Pull request merged**